### DISCUSSION TITLE
Is it possible to automatically read Ansys Mechanical available inputs

### DISCUSSION THREAD

**Discussion by ValentynBarannik** [ansys/pymechanical#449] [Q&A] [answered]
Good day,

I'm truing to find the way automatically how to get inputs (boundary conditions) from Ansys Mechanical to replace the values of them on the new one and perform calculation. the approaches I saw before were based of manual finding the values in ds.dat file. Is there any way to automate this process, like loading project file and recognition will provide me all data (boundary conditions) can be changed for simulation? If yes, could you recommend what file type should be used (*.dat, *.cdb, etc)

Thank you in advance.

**Comment by pmaroneh** [comment]
Hi,
Yes thanks to Mechanical scripting there's an easier way to do that. Open the already existing model (ie, the mechdb file) and then you can use the Mechanical automation API commands to retrieve any object that is present in the tree. See this post for an example: https://discuss.ansys.com/discussion/2504/extract-information-on-boundary-conditions-in-mechanical#latest

**Reply by ValentynBarannik** [reply]
Dear Pernelle,

Thanks for the link. it is exactly what I am looking for. Could you please clarify another think. In the code there is a links to Ansys ACT. where these libraries can be found? should they be installed additionally?

**Reply by pmaroneh** [reply]
Hi Valentyn,
Not sure exactly where you saw a link to Ansys ACT. This term is often used incorrectly, to refer to the Mechanical Automation API, for which you do not need to install anything except the standard Mechanical installation. Ansys ACT stands for "Ansys Customization Toolkit", which enables customizing the UI of Workbench and Mechanical. You don't need to install anything specific either for ACT, except if you want to use an already existing ACT extension, in which case you have to install said extension.

**Reply by ValentynBarannik** [reply]
Hi Pernelle,
I found this link on PyMechanical https://embedding.examples.mechanical.docs.pyansys.com/examples/basic/topology_optimization_cantilever_beam.html#sphx-glr-examples-basic-topology-optimization-cantilever-beam-py

Here, ACT is also used for the presented case. Even if I exclude the loading of ACT library, I have the error with ExtAPI.
Please correct me if I am wrong in my actions. I took Python IDE, put the code from the example (link above) to load *.mechdat, inserted the code you sent, and have the error with ExtAPI.
Can my colleagues have a live meeting with your Team to ask the question?

**Reply by pmaroneh** [reply]
Now I understand to which librairies you were refering to. You do not need any further installation than standard Mechanical and PyMechanical for this example to run. 
I'm afraid having a Teams meeting is not possible as support on PyAnsys is not provided as standard support. However, please create a post in the [Ansys Developer Forum](https://discuss.ansys.com/), share the lines of code you're trying to run, share the exact error message you're getting and we'll try to sort this out.

**Comment by RGPATCHI** [comment]
Hello Valentyn, 

As per our discussion, please have a look at the below at the below post in developer's forum to use PyMechanical to replace boundary conditions based on CSV input. 

https://discuss.ansys.com/discussion/2607/transfer-convection-coefficient-and-ambient-temperature-data-from-csv-to-mechanical-via-pymechanical/p1?new=1

hope this helps.  

If you have any questions, please feel free to comment there.

**Comment by ValentynBarannik** [comment]
Hello Rohith,

Thank you for the example. Our developers informed me that their understanding is much closer to the required. The one nuance we cannot understand. In the example, the value of a variable is replaced by method GetObjectsByName and name is taken from *.csv file. How this *.csv file is prepared? Is it somehow exported from Mechanical project OR User should create this file manually? If 2nd, then how User knows what variables are in the *.mechdat file. In this case we should have complete correspondence between *.csv file and *.mechdat file, and this correspondence , I would say, impossible for manual work.
We assume that we could miss important nuance regarding exporting *.csv from Mechanical project or reading variables from *.mechdat.
Could you please clarify?

Best regards,
Valentyn

**Reply by RGPATCHI** [reply]
Hello Valentyn, 

Thank you for your reply. Please find my feedback below. 

How this *.csv file is prepared? 

I export this csv from Mechanical using scripting. But, please note that, this is just one algorithm (using Names). You can try other algorithms to transfer data between external source (say csv) to Mechanical 

Please have a look at the below post in developers forum to understand how to automate the export
https://discuss.ansys.com/discussion/2611/how-to-export-the-information-of-all-the-convection-loads-in-my-mechanical-to-a-csv/p1?new=1

Best regards,
Rohith

**Reply by ValentynBarannik** [reply]
Hello Rohith,

Thank you for the clarification. The reason of our misunderstanding, and as the result too many question is that we thought that all  functionality is available for API, but it seems you are using another approach like providing the capability to execute any Mechanical related python script inside Mechanical environment.
So, the general approach to work with Mechanical using API is the following: 
1) we create the required Python script that uses ExtAPI. This script can be executed only directly in Mechanical environment, as ExtAPI seems to be the internal method of Mechanical environment.
2) Then, thanks to the bellow code, I can import my script to Mechanical environment and execute it
////////////////////////////////////////////////////////
import os
import ansys.mechanical.core as pymechanical
from ansys.mechanical.core import find_mechanical
 
wb_exe = find_mechanical(232)[0]
 
UserDir = r"D:\Cases\26832"
commandFileName = r"commandFile.py"
commandFilePath = os.path.join(UserDir,commandFileName)
 
mechanical = pymechanical.launch_mechanical(exec_file=wb_exe, batch=False)
print(mechanical)
mechanical.run_python_script_from_file(commandFilePath)
//////////////////////////////////////////////////////////////////////////////////////////////

Am I correct?

Best regards,
Valentyn

**Reply by ValentynBarannik** [reply]
Hello Rohith,

one more question. For some reason I cannot run Python script from IDE, could you please take a look the script
/////////////////////  IDE part///////////////////////////////
import ansys.mechanical.core as pymechanical
from ansys.mechanical.core import find_mechanical

wb_exe = find_mechanical(232)[0]

commandFileName = r"C:\Users\BVS\Desktop\Ansys_Mechanical_integration\read_data.py"

mechanical = pymechanical.launch_mechanical(exec_file=wb_exe, batch=False)
print(mechanical)
mechanical.run_python_script_from_file(commandFileName)
//////////////////////  end of IDE part /////////////////////////////////

///////////////////////////script body/////////////////////////////////
import os

filenameMechDb = r"C:\Users\BVS\Desktop\Ansys_Mechanical_integration\MECHinput.mechdat"
filepath = r"E:"
fileName = 'Convection_Data.csv'

#Provide the name of the CSV file
ExtAPI.DataModel.Project.Open(filenameMechDb)
#Select Analysis by Id
analysis = DataModel.AnalysisList[0]
 
#Function to write a CSV file from the gathered data
def writeCSV(t1,fileName):
    ExtAPI.Log.WriteMessage('my File name ' + fileName)
    with open(fileName , 'w') as f:
        for line in t1:
            for col in line:
                f.write(str(col) + ', ')
            f.write('\n')
 
saveName = os.path.join(filepath,fileName)
first_solution = analysis.Solution
solver_data = first_solution.SolverData
allConvections = DataModel.GetObjectsByType(DataModelObjectCategory.Convection)
allData = []
for convectionObj in allConvections:
    convName = convectionObj.Name
    lastIndexConv = convectionObj.FilmCoefficient.Output.DiscreteValueCount-1
    lastIndexAT = convectionObj.AmbientTemperature.Output.DiscreteValueCount-1
    convFCVal = convectionObj.FilmCoefficient.Output.GetDiscreteValue(lastIndexConv).Value
    convFCUnit = convectionObj.FilmCoefficient.Output.GetDiscreteValue(lastIndexConv).Unit
    convATVal = convectionObj.AmbientTemperature.Output.GetDiscreteValue(lastIndexAT).Value
    convATUnit = convectionObj.AmbientTemperature.Output.GetDiscreteValue(lastIndexAT).Unit
    allData.append([convName, convFCVal, convFCUnit, convATVal, convATUnit])
 
writeCSV(allData,saveName)
 ///////////////////////// end of script /////////////////////////////////////////


Thank you in advance.

BR,
Valentyn

**Reply by RGPATCHI** [reply]
Hello Valentyn, 

Regarding your first question, yes your understanding is correct when using PyMechanical in Remote mode (which is what we used). We can also use PyMechanical in Embedded mode, then you will be able to access the API directly outside Mechanical. But this method has some limitations compared to the Remote method. 

Let me try to modify the script to work with Embedded method. 

Regarding your second question, do you see what the error is? 

I would say, the best option is first to debug the script body (mechanical script) directly inside mechanical by pasting it in the Scripting console to see if its working and debug this further. Then you can automate it from IDE.

**Reply by ValentynBarannik** [reply]
Hello Rohith,

when I use the script inside Mechanical it works correctly. the problem when I use it via API. in this case I'm adding ExtAPI string to say my script what project should be used

/////////////////////////////error code///////////////////////////////////////////////////
Initialize() started
Initialize() done
Ansys Mechanical [Ansys Mechanical Enterprise]
Product Version:232
Software build date: 05/30/2023 15:25:53

CRITICAL -  -  logging - handle_exception - Uncaught exception
Traceback (most recent call last):
  File "C:\Python_prj\Mechanical\main.py", line 11, in <module>
    mechanical.run_python_script_from_file(r"C:\Users\BVS\Desktop\Ansys_Mechanical_integration\read_data.py")
  File "C:\Users\BVS\AppData\Local\Programs\Python\Python312\Lib\site-packages\ansys\mechanical\core\mechanical.py", line 1039, in run_python_script_from_file
    return self.run_python_script(script_code, enable_logging, log_level, progress_interval)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\BVS\AppData\Local\Programs\Python\Python312\Lib\site-packages\ansys\mechanical\core\mechanical.py", line 981, in run_python_script
    result_as_string = self.__call_run_python_script(
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\BVS\AppData\Local\Programs\Python\Python312\Lib\site-packages\ansys\mechanical\core\mechanical.py", line 1702, in __call_run_python_script
    for runscript_response in self._stub.RunPythonScript(request):
  File "C:\Users\BVS\AppData\Local\Programs\Python\Python312\Lib\site-packages\grpc\_channel.py", line 541, in __next__
    return self._next()
           ^^^^^^^^^^^^
  File "C:\Users\BVS\AppData\Local\Programs\Python\Python312\Lib\site-packages\grpc\_channel.py", line 967, in _next
    raise self
grpc._channel._MultiThreadedRendezvous: <_MultiThreadedRendezvous of RPC that terminated with:
	status = StatusCode.UNKNOWN
	details = "Unable to open project. The project is locked by: BVS@VINDEMATRIX at Tue Oct 24 13:35:25 2023
"
	debug_error_string = "UNKNOWN:Error received from peer  {grpc_message:"Unable to open project. The project is locked by: BVS@VINDEMATRIX at Tue Oct 24 13:35:25 2023\n", grpc_status:2, created_time:"2023-10-26T14:02:09.4029068+00:00"}"
>

**Reply by RGPATCHI** [reply]
Hello Valentyn, 

The error indicates that the mechdat file is already open in a previous operation. Before running the script again, please issue the command mechanical.exit() at the end (you can add this command to the end of IDE part. 

Best regards,
Rohith

**Comment by RGPATCHI** [comment, marked as answer]
Hello Valentyn, 

Please find attached a script to do the same task with Embedded method. Also please find in the below discussion on differences between Remote and Embedded sessions. But main difference is that Mechanical APIs are available directly in Python space. 

https://github.com/ansys/pymechanical/discussions/476

```
import os
import logging
import csv
from ansys.mechanical.core import App
from ansys.mechanical.core.embedding.logger import (Configuration,Logger)
Configuration.configure(level=logging.WARNING,to_stdout=True)
app = App(version=232)
print(app)
Logger.error("message")

# Extract the global API entry points (available from built-in Mechanical scripting)
from ansys.mechanical.core import global_variables
# Merge them into your Python global variables
globals().update(global_variables(app))


filenameMechDb = r"MECHinput.mechdat"
filepath = r"D:\Cases\26832"
filepathMechDb = os.path.join(filepath,filenameMechDb)
filenameCSV = r"Convection_Data.csv"
filepathCSV = os.path.join(filepath,filenameCSV)
fileNew = r"MECHinput_modified2.mechdat"
fileNewPath = os.path.join(filepath,fileNew)

ExtAPI.DataModel.Project.Open(filepathMechDb)
  
convData = []
with open(filepathCSV, 'r') as csvfile:
    csv_reader = csv.reader(csvfile)
    for row in csv_reader:
        print(row)
        convName = row[0]
        convFCVal = row[1]
        convFCUnit = row[2]
        convATVal = row[3]
        convATUnit = row[4]
        convData.append([convName, convFCVal, convFCUnit, convATVal, convATUnit])

for i in range(len(convData)):
    ConvObj = DataModel.GetObjectsByName(convData[i][0])[0]
    lastIndexConv = ConvObj.FilmCoefficient.Output.DiscreteValueCount-1
    lastIndexAT = ConvObj.AmbientTemperature.Output.DiscreteValueCount-1
    ConvObj.FilmCoefficient.Output.SetDiscreteValue(lastIndexConv, Quantity(float(convData[i][1]), convData[i][2]))
    ConvObj.AmbientTemperature.Output.SetDiscreteValue(lastIndexConv, Quantity(float(convData[i][3]), convData[i][4]))

ExtAPI.DataModel.Project.Save(fileNewPath)
```

**Reply by ValentynBarannik** [reply]
Hi Rohith,

Unfortunately, the issue with remote approach is not solved. Before, I closed PyMechanical manually. When I execute simple script with code that just writes arbitrary sentence to file, then everything works. But when I use strings with ExtAPI in script, I get error.

By the way, thanks for the embedding code, I will check it

BR,
Valentyn

**Comment by RGPATCHI** [comment]
Hello @ValentynBarannik , 

Is your workflow now working with Embedded approach I proposed?
Please let me know your feedback.

Best regards,
Rohith

**Reply by ValentynBarannik** [reply]
Hello Rohith,

The approach works correctly. Thank you. The problem was in the *.mechdat file. For some reason the file I was using does not work for API, maybe because it was exported with some errors or it is due to export to System disk (despite I do not have any restrictions for System disk) . Once I reexported file, everything starts working correctly. I was able to get the list of boundary conditions.
